### ISSUE TITLE
Database Delete method does not work for derived class

### ISSUE THREAD

**Issue by mergoktas** [CollaboratingPlatypus/PetaPoco#560]
This issue is in V5 but might be continuing in V6 too.

This is my Poco class for a MySql database
```csharp
[TableName("furnace.customerorder")]
[PrimaryKey("ID")]
[ExplicitColumns]
public partial class customerorder : furnaceDB.Record<customerorder>  
{
        [Column] 
	public int ID 
	{ 
		get {  return _ID; }
		set { _ID = value; MarkColumnModified("ID");  }
	}
	int _ID;

       // ... Other columns follow
}
```
and this is another poco class which is derived from above customerorder

```csharp
public  class CustomerOrderPoco : customerorder
{
        [ResultColumn]
        public string CompanyName { get; set; }
        [ResultColumn]
        public string ProcessName { get; set; }
    }
```

And the order object is of type  derived class `CustomerOrderPoco` not the of type `customerorder`
then

```csharp
using (var db = new Database("furnace"))
{
        db.Delete<customerorder>(order);                   
}
```

petapoco producess a MySqlCommand object which has the command text as follow

`DELETE FROM furnace.customerorder WHERE `ID`=@0`

to process the delete request as follow.

Please note that order object has valid value for the ID number as primary key.

Lastly this the method that petapoco executes for deletion

```csharp
public int Execute(string sql, params object[] args)
{
		try
		{
			OpenSharedConnection();
			try
			{
				using (var cmd = CreateCommand(_sharedConnection, sql, args))
				{
					var retv = cmd.ExecuteNonQuery();
					OnExecutedCommand(cmd);
					return retv;
				}
			}
			finally
			{
				CloseSharedConnection();
			}
		}
		catch (Exception x)
		{
			if (OnException(x))
				throw;
			return -1;
		}
    }
```

**Comment by asherber** [comment]
Can you please clarify what the problem is? What isn't working?

**Comment by mergoktas** [comment]
I did the same test again. In PetaPoco.cs  at line 630, I investigated the MySqlCommand object which is cmd variable in source code. Its commant text, as I sad before, as follows
DELETE FROM furnace.customerorder WHERE ID=@0
At this point  shouldn't it be some like ?
DELETE FROM furnace.customerorder WHERE ID=7
I mean the parameter should have been replace with the real ID value.
And after execution , without any problem, again PetaPoco.cs line number 630,
no deletion occurs and retv is equel to zero.

**Comment by asherber** [comment]
The command text looks correct. The parameters should be available in the `MySqlCommand.Parameters` property; they are substituted into the command text by the database server.

What happens when you execute `DELETE FROM furnace.customerorder WHERE ID=7` directly against the database? Is there actually a record with ID=7? I'd like to rule out database issues, because right now it looks like this isn't a PetaPoco issue.

**Comment by mergoktas** [comment]
From the object oriented point of view, I thought that I can use the derived class object instead of main class object. In my case CustomerOrderPoco is derived from customerorder class with some additional [ResultColumn] properties. So I used CustomerOrderPoco  instead of customerorder. 
Besides, When I make cast it from  CustomerOrderPoco to customerorder, VS warns me that cast is redundant. 
In PetaPoco.cs at line 1717, it makes type comparision. Since types are not equal, it skips that line.
` if (pocoOrPrimaryKey.GetType() == typeof(T))
                return Delete(pocoOrPrimaryKey);`

**Comment by mergoktas** [comment]
OK. This is the result directly execution against to database
Query: DELETE FROM furnace.customerorder WHERE ID=7

Error Code: 1451
Cannot delete or update a parent row: a foreign key constraint fails (`furnace`.`workorder`, CONSTRAINT `FK_WorkOrder_CustomerOrder` FOREIGN KEY (`CustomerOrderID`) REFERENCES `customerorder` (`ID`))

Execution Time : 0 sec
Transfer Time  : 0 sec
Total Time     : 0.397 sec

But look at my small subroutine
` private void DeleteCustomerOrder(CustomerOrderPoco order)
        {
            try
            {
                using (var db = new Database("furnace"))
                {
                    db.Delete<customerorder>(order);                   
                }
            }
            catch(Exception ex)
            {
                MyMessageBox.Show(ex.Message);
                return;
            }
              }`
Why not exception is not cought?
I will investigate it more in detail tomorrow and will inform you.
Thanks anyway for your responses today

**Comment by mergoktas** [comment]
I changed my delete routine. Now  I send the main object class, not the object from derived class.
In this case, I can catch the  exception as follow

> Cannot delete or update a parent row: a foreign key constraint fails (furnace.workorder, CONSTRAINT FK_WorkOrder_CustomerOrder FOREIGN KEY (CustomerOrderID) REFERENCES customerorder (ID))

Besides I created another record in database with ID =10.  Nothing has dependent on this record in the database. Delete was successful.

My Conclusion:

> Database Delete method does not work for derived class as in my case following
customerorder is craeted against database table,
CustomerOrderPoco is public drive from customerorder with some additional props,
1.`db.Delete<customerorder>(customerorderObject)` is working, generic parameter indication,
2.`db.Delete(customerorderObject)` is working, no generic parameter indication,
3.`db.Delete(CustomerOrderPocoObject)` is working, no generic parameter indication,
4.`db.Delete<customerorder>(CustomerOrderPocoObject` ) is not, generic parameter indication, 
even does not throw exception. It only returns 0 (nothing deleted) as it is executing normally.

Is 4. one my misunderstanding, or should work normally?
